cpu/sam0_common/periph/timer.c

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (C) 2019 ML!PA Consulting GmbH
+ *
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_sam0_common
+ * @ingroup     drivers_periph_timer
+ * @{
+ *
+ * @file        timer.c
+ * @brief       Low-level timer driver implementation
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "board.h"
+#include "cpu.h"
+
+#include "periph/timer.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief Timer state memory
+ */
+static timer_isr_ctx_t config[TIMER_NUMOF];
+
+static inline TcCount32 *dev(tim_t tim)
+{
+    return &timer_config[tim].dev->COUNT32;
+}
+
+static inline TcCount16 *dev16(tim_t tim)
+{
+    return &timer_config[tim].dev->COUNT16;
+}
+
+static inline TcCount8 *dev8(tim_t tim)
+{
+    return &timer_config[tim].dev->COUNT8;
+}
+
+static inline void wait_synchronization(tim_t tim)
+{
+#if defined(TC_SYNCBUSY_MASK)
+    /* SYNCBUSY is a register */
+    while ((dev(tim)->SYNCBUSY.reg & TC_SYNCBUSY_MASK) != 0) {}
+#elif defined(TC_STATUS_SYNCBUSY)
+    /* SYNCBUSY is a bit */
+    while ((dev(tim)->STATUS.reg & TC_STATUS_SYNCBUSY) != 0) {}
+#else
+#error Unsupported device
+#endif
+}
+
+/* enable timer interrupts */
+static inline void _irq_enable(tim_t tim)
+{
+    NVIC_EnableIRQ(timer_config[tim].irq);
+}
+
+/**
+ * @brief Setup the given timer
+ */
+int timer_init(tim_t tim, unsigned long freq, timer_cb_t cb, void *arg)
```

Comment:
please add a guard in the very beginning of this function to ensure `tim < TIMER_NUMOF`
You can use an assert like other drivers or return an error.

Reply:
Ok.